lightning/src/util/ser_macros.rs

```diff
@@ -40,3 +142,73 @@ macro_rules! impl_writeable_len_match {
 		}
 	}
 }
+
+#[cfg(test)]
+mod tests {
+	use std::io::Cursor;
+	use ln::msgs::DecodeError;
+
+	fn tlv_reader(s: &[u8]) -> Result<(u64, u32, Option<u32>), DecodeError> {
+		let mut s = Cursor::new(s);
+		let mut a: u64 = 0;
+		let mut b: u32 = 0;
+		let mut c: Option<u32> = None;
+		decode_tlv!(&mut s, {(2, a), (3, b)}, {(4, c)});
+		Ok((a, b, c))
+	}
```

Comment:
Hmm, good suggestion (they were added in a later pr, so I missed them). First of all, sadly they can't replace our tests since they don't have any tests for fields which we really want to require (I added a comment noting this).

Second of all, in doing so I found that the original TLV PR isn't the current spec, cause at some point they swapped the endianness of the VarInts (so, great, now we have variable-length integers, which suck, and also a particularly sucky version of variable-length integers, and one that you have to write your own code for and can't re-use the Bitcoin code....WTF?), see https://github.com/lightningnetwork/lightning-rfc/pull/640 so I fixed that while I was at it, as well as tweaked the strictness of our decoder to match the test cases (eg rejecting TLVs that have extra data in a V that we otherwise think we know how to decode).

Finally, this rabbit hole pointed out that somehow we ended up with a second variable-length integer in the spec (?!) unique to TLVs since we already know the length, which I'd also neglected to implement, so that is done as well. Overall it seems like a somewhat nontrivial rewrite, sadly :(.

Reply:
We should submit a PR to the RFC spec to have its test vectors reflect the change in endianness.

Reply:
The RFC spec test vectors do match the current spec, I believe?